client/controller/index.web.js

```diff
@@ -40,6 +42,36 @@ export const ProviderWrappedLayout = ( {
 	const state = store.getState();
 	const userLoggedIn = isUserLoggedIn( state );
 
+	const [ isExperimentLoaded, setIsExperimentLoaded ] = useState( null );
```

Comment:
This runs literally everywhere. At the same time, it's pretty specific and applicable only to part of the application.

IMHO it would be ideal if we could keep experiments for specific workflows away from code that runs everywhere, on every render. 

Just imagine if we add code for dozens more experiments here. I don't think it would be a sustainable way forward.

Is there a really good reason for it to live here? Could we consider moving it somewhere else? A good place could be creating a custom controller method that runs before `makeLayout`. We could then add that only for the routes that actually care about it. 

WDYT?

Reply:
> IMHO it would be ideal if we could keep experiments for specific workflows away from code that runs everywhere, on every render.

Agree, and I would love to do that. 

> We could then add that only for the routes that actually care about it.

I tried to achieve this goal [with this strategy](https://github.com/Automattic/wp-calypso/pull/72932/commits/50a82020b814426d3214e248b67354c27a6687e5#diff-18e8c98551ffdb30f512920db2f0637cbe65e6ac35b2a7d7cbb6164ff81b621eR431) - loading the code only on specific routes. Unfortunately, [I had this issue](https://github.com/Automattic/wp-calypso/pull/72932#issuecomment-1414511295) while trying that approach. Do you know how ew can overcome that @tyxla?

Reply:
Ah, I see. I guess we can't run this on the server for an intentional reason. 

In that case, can we wrap this functionality up as something like a query component then, and drop it only where necessary, instead of loading it for everyone in `<Layout />`?

Reply:
Yeah, I would love to load it only at the [Domain Search component](client/my-sites/domains/domain-search/index.jsx) but the problem we're having is we also need that data available at the Layout, as [we need to decide showing or not](https://github.com/Automattic/wp-calypso/pull/72932/files#diff-7d1eb73ef67c978553b9d47cf17a8f4b9d93bd396ee6cb5ce89e75aedb8b0f19R393) the masterBar and sideBar while mounting the Layout.

Can you point us a better way to load this, making it available for [the Layout menus decision](https://github.com/Automattic/wp-calypso/pull/72932/files#diff-7d1eb73ef67c978553b9d47cf17a8f4b9d93bd396ee6cb5ce89e75aedb8b0f19R388) and in the same time loading only at the Domain and Plans page @tyxla?

Reply:
One way we could do it is by checking if the user is logged in. That's usually a good way to check if we're SSR-ing or not on those isomorphic routes. I see @jessie-ross is also suggesting that [here](https://github.com/Automattic/wp-calypso/pull/72932#discussion_r1098305933).

Note that what you're suggesting is architecturally wrong: blocking the rest of the layout for all screens on Calypso just to check for the value of an experiment that runs only on two sections. This sounds undesirable.

Have we considered trying to provide the experiment value from the server side, as provided by the [current user bootstrap](https://github.com/Automattic/wp-calypso/blob/a14b6727ffc1ed68b951c95f8a43ca6e1cb15ccc/client/server/user-bootstrap/index.js#L28)? That would cause fewer hops and we'll have that data immediately in `<Layout />`. WDYT?

Reply:
Hi, @tyxla 👋🏼 
Thanks for thoughtfully reviewing these bits! Your help is invaluable, as always ❤️ 
I find this scenario a very common one and @Automattic/experimentation-platform is also committed to finding a sustainable approach for future experiments.

> blocking the rest of the layout for all screens on Calypso just to check for the value of an experiment that runs only on two sections

I thought this issue (returning `null` on all screens) is addressed by evaluating `shouldLoadDomainExperiment` (checking for the two sections) in this [condition](https://github.com/Automattic/wp-calypso/pull/72932/files#diff-00073ede4cd5690e0d36147c6bfd9f6e541539a11a8bec5dd8c4d3e36a48ffa2R71).

> Have we considered trying to provide the experiment value from the server side

Yes, the backend approach is detailed by @paulopmt1 in https://github.com/Automattic/wp-calypso/pull/73017, the only caveat being that all Calypso users will be assigned to the experiment instead of doing this based on intent (accessing page `/domains/add` or `/plans/yearly`). If you find it cleaner and more performant, it can work with setting an exposure event (Tracks event) in ExPlat to filter out ineligible users. Then we should definitely document this as best practice 🤓

Reply:
> Yes, the backend approach is detailed by @paulopmt1 in https://github.com/Automattic/wp-calypso/pull/73017, the only caveat being that all Calypso users will be assigned to the experiment instead of doing this based on intent (accessing page /domains/add or /plans/yearly). If you find it cleaner and more performant, it can work with setting an exposure event (Tracks event) in ExPlat to filter out ineligible users. 

That sounds like a very reasonable approach 👍 Any downsides if we decide to go with it from the explat PoV?

> Then we should definitely document this as best practice

💯 Yes, please! This sounds worthy for both Calypso docs and p2.

Reply:
(Sorry there was also a convo happening in slack around this:
p1675758293036799/1675715733.464879-slack-CKZHG0QCR)